edgedb/options.py

```diff
@@ -0,0 +1,291 @@
+            If ``attempts`` is not specified but ``backoff`` is, number of
+            attempts are get from default rule.
+
+            Can't be used if ``options`` is provided.
+
+        :param backoff function:
+            Function that returns amount to sleep between attempts given the
+            iteration number. First retry (which is a second attempt)
+            gets ``n=1``.  If ``attempts`` is not specified but ``backoff``
+            is, number of attempts are get from default rule.
+
+            Make sure to add some randomness to a backoff function.
+
+            Can't be used if ``options`` is provided.
+
+        Examples::
+
+            # Set 4 attempts, but don't change backoff function
+            con1 = con.with_retry(attempts=4)
+
+            # Deadlock with twice delay comparing to default
+            deadlock_fn = lambda n: (2 ** n) * 0.2 + randrange(100) * 0.002
+            # deadlock_fn = lambda n: edgedb.default_backoff(n) * 2
+            con2 = con.with_retry(RetryCondition.Deadlock, backoff=deadlock_fn)
+
+        This method returns a "shallow copy" of the current object
+        with modified transaction options.
+
+        Both ``self`` and returned object can be used after, but when using
+        them transaction options applied will be different.
+
+        Transaction options are are used by both
+        ``raw_transaction`` and ``retrying_transaction``.
+
+        """
+        if options is None:
+            o = self._options.retry_options
+            if condition is None:
+                options = RetryOptions(
+                    attempts=o._attempts if attempts is None else attempts,
+                    backoff=o._backoff if backoff is None else backoff,
+                )
+            else:
+                options = o.with_rule(
+                    condition,
+                    attempts=attempts,
+                    backoff=backoff,
+                )
+        elif (condition is not None or
+              attempts is not None or
+              backoff is not None):
+            raise TypeError("Either options or condition/attempts/backoff "
+                            "should be specified, not both.")
+
+        result = self._shallow_clone()
+        result._options = self._options.with_retry(options)
+        return result
+
+
+class _Options:
+    """Internal class for storing connection options"""
+
+    __slots__ = ['_retry_options', '_transaction_options']
+
+    def __init__(
+        self,
+        retry_options: RetryOptions,
+        transaction_options: TransactionOptions,
+    ):
+        self._retry_options = retry_options
+        self._transaction_options = transaction_options
+
+    @property
+    def retry_options(self):
+        return self._retry_options
+
+    @property
+    def transaction_options(self):
+        return self._transaction_options
+
+    def with_retry_options(self, options: RetryOptions):
```

Comment:
I don't see tests for this method. But as I mentioned in another comment, the need to create `RetryOptions` object separately makes the API extremely verbose and make it feel low level.

Reply:
This is not a user-visible method. So it get's normalized `RetryOptions` object.

Reply:
Well, let's move my comment to line 172 ;) The question still stands: `_OptionsMixin.with_retry_options` seems to lack tests, no?

Reply:
Yes, no tests options setters. I'll add when we decide that current approach is okay.

Reply:
Added tests for retry with different options.